### PR TITLE
[extractor/TwitCasting] fix: Use UserNotLive

### PR DESCRIPTION
**Pull request by at-wat**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

Running yt-dlp with `--wait-for-video` option for TwitCastingLive URL results an error:
```shell
$ yt-dlp --wait-for-video=30 https://twitcasting.tv/channel_name
[TwitCastingLive] Extracting URL: https://twitcasting.tv/channel_name
[TwitCastingLive] Downloading live video of user channel_name. Pass "https://twitcasting.tv/channel_name/show" to download the history
[TwitCastingLive] channel_name: Downloading webpage
[TwitCastingLive] channel_name: Downloading live history
ERROR: [TwitCastingLive] channel_name: The user is not currently live; please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
$
```

This is caused since `class TwitCastingLiveIE` raises `ExtractorError` instead of `UserNotLive` when the channel is currently not live.

This PR fixes `TwitCastingIE` to raise `UserNotLive` when the channel isn't live.
`--wait-for-video` option for TwitCasting works with this branch as below:
```
$ yt-dlp --wait-for-video=30 https://twitcasting.tv/channel_name
[TwitCastingLive] Extracting URL: https://twitcasting.tv/channel_name
[TwitCastingLive] Downloading live video of user channel_name. Pass "https://twitcasting.tv/channel_name/show" to download the history
[TwitCastingLive] channel_name: Downloading webpage
[TwitCastingLive] channel_name: Downloading live history
WARNING: [TwitCastingLive] channel_name: The channel is not currently live
[wait] Waiting for 00:00:30 - Press Ctrl+C to try now
[wait] Remaining time until next attempt: 00:00:22
```

Added a test case to check the raised exception against a dummy TwitCasting account which won't be live.
```shell
$ python test/test_download.py TestDownload.test_TwitCastingLive_all
[debug] Loaded 1864 extractors
[TwitCastingLive] Extracting URL: https://twitcasting.tv/c:unusedlive
[TwitCastingLive] Downloading live video of user c:unusedlive. Pass "https://twitcasting.tv/c:unusedlive/show" to download the history
[TwitCastingLive] c:unusedlive: Downloading webpage
[TwitCastingLive] c:unusedlive: Downloading live history
ERROR: [TwitCastingLive] c:unusedlive: The channel is not currently live
  File "/home/at-wat/Downloads/yt-dlp/yt_dlp/extractor/common.py", line 715, in extract
    ie_result = self._real_extract(url)
                ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/at-wat/Downloads/yt-dlp/yt_dlp/extractor/twitcasting.py", line 267, in _real_extract
    raise UserNotLive(video_id=uploader_id)

.
----------------------------------------------------------------------
Ran 1 test in 0.487s

OK
$ 
```

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details><details><summary>Copilot Summary</summary>  

### Summary
🧪🚨🎥

This pull request enhances the testing and error handling of `test_download` and `twitcasting` modules. It adds support for testing expected exceptions in `test_download` and raises a specific exception for non-live streams in `twitcasting`.

> _`YoutubeDLError` is the sign of doom_
> _We skip the tests that seal our fate_
> _`UserNotLive` is the curse we loom_
> _We raise the exception and test our hate_

### Walkthrough
* Import `YoutubeDLError` and `UserNotLive` exceptions from `yt_dlp.utils` ([link](https://github.com/yt-dlp/yt-dlp/pull/7975/files?diff=unified&w=0#diff-75b6f51c83b83cd3d4b1e27953d72026ddb983ef7122fae0d64ebcba5334f537R34), [link](https://github.com/yt-dlp/yt-dlp/pull/7975/files?diff=unified&w=0#diff-da40a341858d2a623c9ce4832d7be0117985c613125ccc6184b3e078f8251dd0R10))
* Add and modify test cases for `TwitCastingLiveIE` extractor to expect `UserNotLive` exception when user is not live ([link](https://github.com/yt-dlp/yt-dlp/pull/7975/files?diff=unified&w=0#diff-da40a341858d2a623c9ce4832d7be0117985c613125ccc6184b3e078f8251dd0R239-R241), [link](https://github.com/yt-dlp/yt-dlp/pull/7975/files?diff=unified&w=0#diff-75b6f51c83b83cd3d4b1e27953d72026ddb983ef7122fae0d64ebcba5334f537R104-R105))
* Raise `UserNotLive` exception in `TwitCastingLiveIE._real_extract` method when `current_live` value is missing from webpage ([link](https://github.com/yt-dlp/yt-dlp/pull/7975/files?diff=unified&w=0#diff-da40a341858d2a623c9ce4832d7be0117985c613125ccc6184b3e078f8251dd0L263-R267))
* Add helper method to check if raised exception matches expected exception in test function ([link](https://github.com/yt-dlp/yt-dlp/pull/7975/files?diff=unified&w=0#diff-75b6f51c83b83cd3d4b1e27953d72026ddb983ef7122fae0d64ebcba5334f537R145-R155))
* Call helper method in `except` blocks and return or re-raise exception accordingly in test function ([link](https://github.com/yt-dlp/yt-dlp/pull/7975/files?diff=unified&w=0#diff-75b6f51c83b83cd3d4b1e27953d72026ddb983ef7122fae0d64ebcba5334f537R178-R179), [link](https://github.com/yt-dlp/yt-dlp/pull/7975/files?diff=unified&w=0#diff-75b6f51c83b83cd3d4b1e27953d72026ddb983ef7122fae0d64ebcba5334f537R190-R193))

</details>

### CODE REVIEW

**Review by pukkandan — APPROVED** [review]
LGTM. But the core and extractor changes should be merged in separate commits

File: test/test_download.py [hunk 1]
```diff
@@ -139,6 +142,17 @@ def get_tc_filename(tc):
 
         res_dict = None
 
+        def match_exception(self, err):
```
Comment (bashonly) [inline test/test_download.py:145]:
```suggestion
        def match_exception(err):
```

File: test/test_download.py [hunk 2]
```diff
@@ -161,6 +175,8 @@ def try_rm_tcs_files(tcs=None):
                 except (DownloadError, ExtractorError) as err:
                     # Check if the exception is not a network related one
                     if not isinstance(err.exc_info[1], (TransportError, UnavailableVideoError)) or (isinstance(err.exc_info[1], HTTPError) and err.exc_info[1].status == 503):
+                        if match_exception(self, err):
```
Comment (bashonly) [inline test/test_download.py:178]:
```suggestion
                        if match_exception(err):
```

File: test/test_download.py [hunk 3]
```diff
@@ -171,6 +187,10 @@ def try_rm_tcs_files(tcs=None):
                     print(f'Retrying: {try_num} failed tries\n\n##########\n\n')
 
                     try_num += 1
+                except YoutubeDLError as err:
+                    if match_exception(self, err):
```
Comment (bashonly) [inline test/test_download.py:191]:
```suggestion
                    if match_exception(err):
```

File: test/test_download.py [hunk 4]
```diff
@@ -100,6 +101,8 @@ def print_skipping(reason):
             print_skipping('IE marked as not _WORKING')
 
         for tc in test_cases:
+            if tc.get('expected_exception', None):
```
Comment (bashonly) [inline test/test_download.py:104]:
```suggestion
            if tc.get('expected_exception'):
```

**Review by bashonly — APPROVED** [review]

File: yt_dlp/extractor/twitcasting.py [hunk 5]
```diff
@@ -7,6 +7,7 @@
 from ..utils import (
     clean_html,
     ExtractorError,
+    UserNotLive,
```
Comment (bashonly) [inline yt_dlp/extractor/twitcasting.py:10]:
```suggestion
    ExtractorError,
    UserNotLive,
    clean_html,
```

File: test/test_download.py [hunk 6]
```diff
@@ -139,6 +142,17 @@ def get_tc_filename(tc):
 
         res_dict = None
 
+        def match_exception(err):
+            expected_exception = test_case.get('expected_exception', None)
```
Comment (bashonly) [inline test/test_download.py:146]:
```suggestion
            expected_exception = test_case.get('expected_exception')
```

**Pull request merged**